---
title: "Plots Revised"
author: "Evan Gorstein"
date: "2023-12-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggh4x)
library(patchwork)
library(RColorBrewer)
library(latex2exp)
library(kableExtra)
library(directlabels)
# set plotting theme -----------
theme_set(theme_bw())
theme_update(panel.grid = element_blank())

my_colors = c("orange", "royalblue")
scale_color_mine = function(...) {
  ggplot2:::manual_scale("color", 
               values = setNames(my_colors, c("LASSO", "SCAD")),
               ...)
}

otu_colors = c("#43CD7F", "#9B30FF")
scale_color_otu = function(...) {
  ggplot2:::manual_scale("color", 
               values = setNames(otu_colors, c("band", "scale_free")),
               ...)
}

corr.labs = c("yes correlation", "no correlation")
names(corr.labs) <- c("Yes", "No")

```

# Read in the data:
```{r, message=FALSE}
gene_results <- read_csv("../sim_results/gene/best_results.csv") 
gwas_results <- read_csv("../sim_results/gwas/best_results.csv")
otu_results <- read_csv("../sim_results/OTU/best_results.csv")
```

These data frames contain the fits with the best two BIC's (across all lambdas) for each dataset generated under each setting.



# Gene expression results
Let's check on the count for each setting

```{r}
gene_results %>%
  count(setting)
```

These numbers should all be 200. Unfortunately, they're not all 200 because for some settings not all 100 data sets receive top two results because some datasets failed to converge for all lambdas.

Looking at just the best lambda: 

```{r}
best_gene_results <- gene_results %>% 
  arrange(setting, data_id, bic) %>% 
  group_by(setting, data_id) %>%
  filter(row_number() == 1) %>%
  ungroup()

count(best_gene_results, setting)
```


These numbers should all be 100. 


Closer look to see what's missing

```{r}


bgr_filt <- best_gene_results %>%
  group_by(setting) %>%
  mutate(n = n()) %>%
  ungroup() %>%
  filter(n < 100) %>%
  select(setting, data_id, lambda, n_nz, sigma, psi_1, bic) %>%
  arrange(setting, data_id) 
all <- expand(bgr_filt, setting, data_id)

#Missing
all %>% anti_join(bgr_filt)

#Missing appear at the end of this data frame with NAs in all but first 2 columns
bgr_filt %>% 
  right_join(all)

```

## Variable selection 

### Main text 

We want to plot the distribution of false and true positives for the simulations with 10 true fixed effects, 3 of which are random effects.

First, we filter our `best_gene_results` data-frame to only those simulations that had 10 true fixed effects:
```{r}
bgr_nz10 <- best_gene_results %>%
  filter(str_detect(setting, "nz10")) %>%
  mutate(
    cor = str_extract(setting, "rho([0-9]+\\.[0-9]+)", group = 1),
    penalty = setting %>% 
      str_extract("cov.*_(.*)-", group = 1) %>% toupper %>% factor,
    q = str_extract(setting, "random([0-9])", group = 1),
    cov_str = recode_factor(
      str_extract(setting, "cov(.*)_", group = 1),
      id = "scalar", diag = "diagonal", sym = "unstructured"
    ),
    predictor_cor = if_else(cor == "0.0", "No", "Yes"),
    fp = n_nz - tp,
    f = (2*tp)/(tp+fp+10)
  ) 

```


Now we can create the plot of false positives for those simulations for which there were 3 random effects:

```{r}
gene_main_fp <- bgr_nz10 %>%
  filter(q == 3) %>%
  mutate(fpr = fp/990) %>%
ggplot(aes(x=1, y = fpr, col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(pch=21, position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2)) +
  labs(y = "false positive rate") +
  facet_grid2(predictor_cor ~ cov_str, scales = "free_y", independent = "y",
             labeller = labeller(predictor_cor = corr.labs)) +
  #scale_y_continuous(expand = expansion(mult = c(0, .05))) +
  theme(legend.position = "top",
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gene_main_fp

#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gene/gene_main_fp.pdf",
  gene_main_fp,
  width = 20,
  height = 15,
  units = "cm"
)
```

We can similarly create the plot of true positives for those simulations for which there were 3 random effects:

```{r}
# Export the data for this plot to a csv for Rosa to make a table

bgr_nz10 %>%
  filter(q==3) %>%
  select(penalty, cov_str, predictor_cor, tp) %>%
  write_csv("../sim_results/true_positive_tables/gene/gene_main_tp.csv")


```


```{r, warning=FALSE}

counts <- bgr_nz10 %>%
  group_by(penalty, cov_str, q, predictor_cor) %>%
  summarise(
    cnt_perfect = sum(tp == 10), .groups = "drop",
    ) %>% 
  filter(q == 3)

gene_main_tp <- bgr_nz10 %>%
  filter(q==3) %>%
  filter(tp < 10) %>%
  bind_rows(
    data.frame(
      tp = c(NA, NA), penalty = c("SCAD", "LASSO"), 
      predictor_cor = c("No", "No"), 
      cov_str = factor(c("scalar", "unstructured"), levels = c("scalar", "diagonal", "unstructured"))
    )) %>%
  ggplot(aes(y = tp, x = cov_str, col = penalty)) +
  geom_point(
    pch = 21,
    position = position_jitterdodge(
      dodge.width = 0.9,
      jitter.height = 0,
      jitter.width = 0.25)
    ) +
  geom_text(data = counts, aes(cov_str, 10, label = cnt_perfect),
            show.legend = FALSE,
            position = position_dodge(width = 0.9), size = 5) +
  labs(y = "# true positives (out of 10)", x = TeX("$\\Psi_\\theta$ structure")) +
  scale_y_continuous(breaks = c(3,6,9,10), expand = expansion(mult = c(.05, .1))) +
  facet_wrap(~ predictor_cor, nrow = 2, labeller = labeller(predictor_cor = corr.labs)) +
  theme(
    legend.position = "top",
    axis.ticks.x = element_blank(),
    text = element_text(size = 18), 
    axis.title.y = element_text(size = rel(.85))
        ) +
  scale_color_mine()

gene_main_tp

ggsave( 
  "../plots/gene/gene_main_tp.pdf",
  gene_main_tp,
  width = 20,
  height = 15,
  units = "cm"
) 
```

Let's combine these using patchworks:
```{r, warning=FALSE}
gene_main_fp + gene_main_tp + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gene/gene_main_select.pdf", width = 40, height = 15, units = "cm")
```


### Supplementary material
First, we filter our `best_gene_results` data-frame to only those simulations that had 5 true fixed effects:
```{r}
bgr_nz5 <- best_gene_results %>%
  filter(str_detect(setting, "nz5")) %>%
  mutate(
    cor = str_extract(setting, "rho([0-9]+\\.[0-9]+)", group = 1),
    penalty = setting %>% 
      str_extract("cov.*_(.*)-", group = 1) %>% toupper %>% factor,
    p = as.numeric(
      str_extract(setting, "dim([0-9]+)", group = 1)
    ),
    predictor_cor = if_else(cor == "0.0", "No", "Yes"),
    fp = n_nz - tp,
    f = (2*tp)/(tp+fp+5)
  ) 
```

Now we can create the plot of false positives for those simulations that had 5 true fixed effects:

```{r}
gene_supp_nz5_fp <- bgr_nz5 %>% 
  group_by(penalty, p, predictor_cor) %>%
  mutate(outlier = fp > quantile(fp, .75) + IQR(fp) * 1.5) %>%
  ungroup() %>% 
  ggplot(aes(as.factor(p), fp/(p-5), col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), outlier.shape = NA) +
  geom_point(data = function(x) dplyr::filter(x, outlier), pch=21, 
              position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.3)) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
  labs(y = "false positive rate", x = TeX("dimension ($p$)")) +
  theme(
    legend.position = "top",
    axis.ticks.x = element_blank(),
    text = element_text(size = 18), 
    axis.title.y = element_text(size = rel(.85))
        ) +
  scale_color_mine()

gene_supp_nz5_fp
ggsave(
    "../plots/gene/gene_supp_nz5_fp.pdf",
    gene_supp_nz5_fp,
    width = 15,
    height = 15,
    units = "cm"
  )

```


We can similarly create the plot of true positives for those simulations for which there were 5 true fixed effects:
```{r}

counts <- bgr_nz5 %>%
  group_by(penalty, p, predictor_cor) %>%
  summarise(
    cnt_perfect = sum(tp == 5), .groups = "drop",
    ) 
gene_supp_nz5_tp <- bgr_nz5 %>% 
  filter( tp < 5) %>%
  ggplot(aes(factor(p), tp, col = penalty)) +
  geom_point(pch=21, 
              position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.3)) +
  geom_text(data = counts, aes(factor(p), 5, label = cnt_perfect),
            show.legend = FALSE,
            position = position_dodge(width = 0.9), size = 5) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
  labs(y = "# true positives (out of 5)", x = TeX("dimension ($p$)")) +
  scale_y_continuous(limits = c(0,5),  expand = expansion(mult = c(.05, .1))) +
  theme(
    legend.position = "top",
    axis.ticks.x = element_blank(),
    text = element_text(size = 18), 
    axis.title.y = element_text(size = rel(.85))
        ) +
  scale_color_mine()

gene_supp_nz5_tp
ggsave(
    "../plots/gene/gene_supp_nz5_tp.pdf",
    gene_supp_nz5_tp,
    width = 15,
    height = 15,
    units = "cm"
  )

```

We can create the plot of false positives for those simulations that had 10 true fixed effects, and 1 or 5 random effects.
```{r, warning=FALSE}
no_out = data.frame(q = c("1", "5"), 
                    penalty = c("LASSO", "LASSO"),
                    predictor_cor = c("No", "Yes"), 
                    fp = NA,
                    outlier = TRUE,
                    plot = FALSE)

gene_supp_nz10_fp <- bgr_nz10 %>%
  filter(q != 3) %>%
  group_by(penalty, q, predictor_cor) %>%
  mutate(outlier = fp > quantile(fp, .75) + IQR(fp) * 1.5,
         plot = TRUE) %>%
  ungroup() %>% 
  bind_rows(no_out) %>%
  {
  ggplot(data =., aes(y = fp/990, x = q, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2)) +
  labs(y = "false positive rate", x = TeX("# random effects ($q$)")) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
}
gene_supp_nz10_fp
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gene/gene_supp_nz10_fp.pdf",
  gene_supp_nz10_fp,
  width = 15,
  height = 15,
  units = "cm"
)
```

We can similarly create the plot of true positives for those simulations that had 10 true fixed effects, and 1 or 5 random effects.
```{r, warning=FALSE}

counts <- bgr_nz10 %>%
  group_by(penalty, q, predictor_cor) %>%
  summarise(
    cnt_perfect = sum(tp == 10), .groups = "drop",
    ) %>%
  filter(q!=3)
no_out = data.frame(q = c("1", "1", "5"), 
                    penalty = c("SCAD", "SCAD", "LASSO"),
                    predictor_cor = c("No", "Yes", "No"), 
                    tp = NA)

gene_supp_nz10_tp <- bgr_nz10 %>%
  filter(q != 3) %>%
  filter(tp < 10) %>%
  bind_rows(no_out) %>%
  ggplot(aes(y = tp, x = q, 
                      col = penalty)) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2)) +
  geom_text(data = counts, aes(q, 10, label = cnt_perfect),
            show.legend = FALSE,
            position = position_dodge(width = 0.9), size = 5) +
  labs(y = "# true positives (out of 10)", x = TeX("# random effects ($q$)")) +
  facet_wrap(~ predictor_cor, nrow = 2, labeller = labeller(predictor_cor = corr.labs)) +
  theme(
    legend.position = "top",
    axis.ticks.x = element_blank(),
    text = element_text(size = 18), 
    axis.title.y = element_text(size = rel(.85))
        ) +
  scale_y_continuous(breaks = c(1,5,9,10), 
                     labels = scales::label_number(accuracy=1),
                     expand = expansion(mult = c(.05, .1))
                     ) +
  scale_color_mine()

gene_supp_nz10_tp
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gene/gene_supp_nz10_tp.pdf",
  gene_supp_nz10_tp,
  width = 15,
  height = 15,
  units = "cm"
)
```

Now we can combine the above 4 plots with patchwork:
```{r, warning=FALSE}
(gene_supp_nz5_fp + gene_supp_nz5_tp) / (gene_supp_nz10_fp + gene_supp_nz10_tp) + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gene/gene_supp_select.pdf", width = 40, height = 30, units = "cm")
```




### F-scores

Here's a table summary of f-scores
```{r}
bgr_nz5 %>%
  group_by(penalty, p, predictor_cor) %>% 
  summarise(mn = mean(f), min = min(f), med = median(f), max = max(f), n = n())
bgr_nz10 %>%
  group_by(penalty, cov_str, q, predictor_cor) %>% 
  summarise(mn = mean(f), min = min(f), med = median(f), max = max(f), n = n())
```

Here's a table summary of false positives 
```{r}
bgr_nz5 %>%
  group_by(penalty, p, predictor_cor) %>% 
  summarise(mn = mean(fp), min = min(fp), med = median(fp), max = max(fp), n = n())
bgr_nz10 %>%
  group_by(penalty, cov_str, q, predictor_cor) %>% 
  summarise(mn = mean(fp), min = min(fp), med = median(fp), max = max(fp), n = n())
```


## Estimation


### Main text

We want to plot the distribution of a penalized estimator and the distribution of an unpenalized estimator for the simulations with 10 true fixed effects, 3 of which are random effects.


We start with an unpenalized estimator. Let's look at $\hat{\beta_2}$, which was unpenalized because $q=3$. 
```{r}
df_true <- data.frame(y_value = 2) 

gene_main_beta2 <- bgr_nz10 %>%
  filter(q == 3) %>%
  ggplot(aes(y = beta_2, x = cov_str, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(pch=21, position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2)) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_2$"), x = TeX("$\\Psi_\\theta$ structure")) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gene_main_beta2
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gene/gene_main_beta2.pdf",
  gene_main_beta2,
  width = 20,
  height = 15,
  units = "cm"
)
```

What about $\beta_3$? Weirdly, we get bias in the LASSO estimator for settings with correlated features...even though there isn't anything really different about $\beta_3$ than $\beta_2$!
```{r}
df_true <- data.frame(y_value = 4) 

gene_main_beta3 <- bgr_nz10 %>%
  filter(q == 3) %>%
  ggplot(aes(y = beta_3, x = cov_str, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2))+
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  scale_y_continuous(breaks = c(3,4,5,6,7,8,9), labels = scales::label_number(accuracy=1))+
  labs(y = TeX("$\\beta_3$"), x = TeX("$\\Psi_\\theta$ structure")) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs), scales = "free") +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
gene_main_beta3
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gene/gene_main_beta3.pdf",
  gene_main_beta3,
  width = 20,
  height = 15,
  units = "cm"
)
```
Remember, bias towards 0 is expected only for coefficients that have a penalty applied for the LASSO (and $\beta_3$, just like $\beta_2$ is unpenalized). There seem to be two groups of estimates with both clusters biased but one more than the other. And the bias is worse for the scalar covariance structure. This likely has to do with the fact that the LASSO is trying to use $\beta3$ to compensate for setting $\beta_4:\beta_10$ to 0 in these settings (see the plot of true positives).

Now, let's take a look at a penalized parameter, $\beta_4$.

```{r}
counts <- bgr_nz10 %>%
  filter(q == 3) %>%
  filter(beta_4 == 0) %>%
  group_by(predictor_cor, cov_str, penalty) %>%
  summarise( n = n(), .groups = "drop") %>%
  complete(predictor_cor, cov_str, penalty, fill = list(n = 0)) 

df_true <- data.frame(y_value = 3) 

gene_main_beta4 <- bgr_nz10 %>%
  filter(q == 3) %>%
  filter(beta_4 != 0) %>%
  group_by(penalty, cov_str, predictor_cor) %>%
  ungroup() %>% 
  {
  ggplot(data =., aes(y = beta_4, x = cov_str, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), outlier.shape = NA) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2))+
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  geom_text(data = counts, aes(cov_str, 0, label = n), 
            position = position_dodge(width = 0.9), size = 5, 
            show.legend  = F) +
  labs(y = TeX("$\\beta_4$"), x = TeX("$\\Psi_\\theta$ structure")) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
  scale_y_continuous(limits = c(-0.5, 3.5), labels = scales::label_number(accuracy=1)) +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
  }

gene_main_beta4
ggsave(
  "../plots/gene/gene_main_beta4.pdf",
  gene_main_beta4,
  width = 20,
  height = 15,
  units = "cm"
)
```
 
Now we can combine the above 2 plots with patchwork:
```{r warning=FALSE}
gene_main_beta3 + gene_main_beta4 + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gene/gene_main_est.pdf", width = 40, height = 15, units = "cm")
```
Check relationship between beta_3 and beta_4 for Lasso with correlation

```{r}
compensation <- bgr_nz10 %>%
  filter(q == 3, penalty == "LASSO", predictor_cor == "Yes") %>%
  ggplot(aes(x = beta_4, y = beta_3, col=penalty)) +
  geom_hline(yintercept = 4)+
  geom_vline(xintercept = 3)+
  geom_point(pch = 21)+
  facet_wrap(~cov_str) +
  labs(y = TeX("$\\beta_3$"), x = TeX("$\\beta_4$")) +
  scale_color_mine()

compensation
ggsave(
  "../plots/gene/compensation.pdf",
  compensation,
  width = 20,
  height = 15,
  units = "cm"
)
```
```{r}
compensation_7 <- bgr_nz10 %>%
  filter(q == 3, penalty == "LASSO") %>%
  ggplot(aes(x = beta_4, y = beta_3, col=penalty)) +
  geom_hline(yintercept = 4)+
  geom_vline(xintercept = 3)+
  geom_point(pch = 21)+
  facet_grid(predictor_cor~cov_str) +
  labs(y = TeX("$\\beta_3$"), x = TeX("$\\beta_4$")) +
  scale_color_mine()

compensation_7
```

```{r}
bgr_nz10 %>%
  filter(q == 3, penalty == "LASSO", predictor_cor == "Yes") %>%
  filter(beta_10 != 0) %>%
  group_by(cov_str) %>%
  summarise(cor(beta_3,beta_4))
```

Let's create a plot that shows parameter estimates for all non-zero coefficients. We do this for a specific setting, the setting with correlated features and 3 random effects with unstructured covariance, at once

```{r}
df_long <- bgr_nz10 %>%
  filter(q == 3) %>%
  pivot_longer(cols = num_range("beta_", 1:10), names_to = "coefficient", values_to = "estimate") %>%
  select(data_id, penalty, coefficient, predictor_cor, cov_str, estimate) %>%
  mutate(coefficient = factor(coefficient, levels = paste0("beta_", 1:10)))

df_true <- data.frame(
  coefficient = factor(paste0("beta_", 1:10), levels = paste0("beta_", 1:10)), 
  y_value = c(1,2,4,3,3,-1,5,-3,2,2)
)
```

```{r}
df_long <- filter(df_long,
                  cov_str == "unstructured", 
                  predictor_cor == "Yes")
all_coefs <- ggplot(df_long, aes(x = coefficient, y = estimate)) +
  geom_boxplot(aes(col = penalty),
               position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(aes(col = penalty),
             pch = 21,
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2)) +
  geom_segment(data = df_true, 
               aes(x = as.numeric(coefficient) - .4, 
                   xend = as.numeric(coefficient) + .4,
                   y = y_value, yend = y_value,
                   linetype = "True Parameter"),
               color = "black",
               show.legend = FALSE) +
   # facet_grid(cov_str ~ predictor_cor, 
   #           labeller = labeller(predictor_cor = corr.labs)) +
  scale_x_discrete(labels = c( TeX("$\\beta_{1}"),
                               TeX("$\\beta_{2}"),
                               TeX("$\\beta_{3}"),
                               TeX("$\\beta_{4}"),
                               TeX("$\\beta_{5}"),
                               TeX("$\\beta_{6}"),
                               TeX("$\\beta_{7}"),
                               TeX("$\\beta_{8}"),
                               TeX("$\\beta_{9}"),
                               TeX("$\\beta_{10}"))) +
  scale_y_continuous(limits = c(-3.5,NA), 
                     labels = scales::label_number(accuracy=1),
                     n.breaks = 6) +
  theme(legend.position = "top",
        text = element_text(size = 18)) +
  scale_color_mine()

all_coefs
ggsave(
  "../plots/gene/all_coefs.pdf",
  all_coefs,
  width = 20,
  height = 15,
  units = "cm"
)
```

Let's get MC means and sds of beta_8 estimates under SCAD penalty:

```{r}
gene_beta8 <- bgr_nz10 %>%
  filter(penalty == "SCAD") %>%
  select(setting, data_id, beta_8) %>%
  mutate(identified = beta_8!=0) %>%
  filter(identified) %>%
  group_by(setting) %>%
  summarise(recovered = n(), mean = mean(beta_8), 
            bias = mean + 3, sd = sd(beta_8))

gene_beta8
```





### Supplementary material

Distribution of $beta_3$ in settings with 1 and 5 random effects. In the first, $beta_3$ is penalized; in the latter, it is unpenalized.
```{r, warning=FALSE}
df_true <- data.frame(y_value = 4) 
counts <- bgr_nz10 %>%
  filter(q != "3") %>%
  filter(beta_3 == 0) %>%
  group_by(predictor_cor, q, penalty) %>%
  summarise( n = n(), .groups = "drop") %>%
  complete(predictor_cor, q=c("1","5"), penalty, fill = list(n = 0))

gene_supp_beta3 <- bgr_nz10 %>%
  filter(q != "3") %>%
  filter(beta_3 != 0) %>%
  group_by(penalty, cov_str, predictor_cor) %>%
  ungroup() %>% 
  {
    ggplot(data =., aes(y = beta_3, x = q, 
                      col = penalty)) +
      geom_boxplot(position = position_dodge(width = 0.9), outlier.shape = 21) +
      geom_text(data = counts, aes(q, 0, label = n), 
            position = position_dodge(width = 0.9), size = 5, 
            show.legend  = F) +
      geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
      labs(y = TeX("$\\beta_3$"), x = TeX("# random effects ($q$)")) +
      scale_y_continuous(limits = c(-0.5, 4.5), labels = scales::label_number(accuracy=1)) +
      facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
      theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
      scale_color_mine()
}
gene_supp_beta3
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gene/gene_supp_beta3.pdf",
  gene_supp_beta3,
  width = 15,
  height = 15,
  units = "cm"
)
```

Distribution of penalized $beta_4$ in settings with 1 and 5 random effects.
```{r}
counts <- bgr_nz10 %>%
  filter(q != "3") %>%
  filter(beta_4 == 0) %>%
  group_by(predictor_cor, q, penalty) %>%
  summarise( n = n(), .groups = "drop") %>%
  complete(predictor_cor, q=c("1","5"), penalty, fill = list(n = 0)) 

df_true <- data.frame(y_value = 3) 

gene_supp_beta4 <- bgr_nz10 %>%
  filter(q != 3) %>%
  filter(beta_4 != 0) %>%
  {
  ggplot(data =., aes(y = beta_4, x = q, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), outlier.shape = 21) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  geom_text(data = counts, aes(q, 0, label = n), 
            position = position_dodge(width = 0.9), size = 5, 
            show.legend  = F) +
  labs(y = TeX("$\\beta_4$"), x = TeX("# random effects ($q$)")) +
  facet_wrap(~ predictor_cor, nrow = 2, 
             labeller = labeller(predictor_cor = corr.labs)) +
  scale_y_continuous(limits = c(-.5, 3.5), labels = scales::label_number(accuracy=1)) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
  }

gene_supp_beta4
ggsave(
  "../plots/gene/gene_main_beta4.pdf",
  gene_supp_beta4,
  width = 15,
  height = 15,
  units = "cm"
)
```


Distribution of penalized $beta_3$ in settings with only 5 non-zero fixed effects

```{r, warning=FALSE}
df_true <- data.frame(y_value = 4) 
gene_supp_nz5_beta3 <- bgr_nz5 %>% 
  ggplot(aes(y = beta_3, x = factor(p), 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), outlier.shape = NA) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2),
             show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_3$"), x = TeX("dimension ($p$)")) +
  facet_wrap(~ predictor_cor, nrow = 2,
             labeller = labeller(predictor_cor = corr.labs)) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gene_supp_nz5_beta3
ggsave(
    "../plots/gene/gene_supp_nz5_beta3.pdf",
    gene_supp_nz5_beta3,
    width = 15,
    height = 15,
    units = "cm"
  )
```

```{r}
df_true <- data.frame(y_value = 3) 
gene_supp_nz5_beta4 <- bgr_nz5 %>% 
  ggplot(aes(y = beta_4, x = factor(p), 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), outlier.shape = NA) +
  geom_point(
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2),
             show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_4$"), x = TeX("dimension ($p$)")) +
  facet_wrap(~ predictor_cor, nrow = 2,
             labeller = labeller(predictor_cor = corr.labs)) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gene_supp_nz5_beta4
ggsave(
    "../plots/gene/gene_supp_nz5_beta4.pdf",
    gene_supp_nz5_beta4,
    width = 15,
    height = 15,
    units = "cm"
  )
```



Now we can combine the above 4 plots for our supplementary materials estimation figure:

```{r}
(gene_supp_nz5_beta3 + gene_supp_nz5_beta4) / (gene_supp_beta3 + gene_supp_beta4) + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gene/gene_supp_est.pdf", width = 40, height = 40, units = "cm")
```

## Variance components

### Scalar

Variance components when covariance structure is scalar
```{r}

df_true <- data.frame(y_value = 0.56) 

var_ge_scalar <- bgr_nz10 %>%
  filter(cov_str == "scalar") %>%
  ggplot(aes(y = psi_1, x=q)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21)  +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = "Estimated parameter", x = TeX("# random effects ($q$)")) +
  facet_wrap(~ predictor_cor, nrow = 2,
             labeller = labeller(predictor_cor = corr.labs)) +
  scale_y_continuous(labels = scales::label_number(accuracy=.1)) +
  coord_cartesian(ylim = c(0,1)) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
  
var_ge_scalar


ggsave(
  "../plots/gene/var_ge_scalar.pdf",
  var_ge_scalar,
  width = 20,
  height = 15,
  units = "cm"
)
```
### Diagonal

```{r}
df_true <- data.frame(parameter = rep(c(1, 2, 3), 2), 
                      y_value = rep(c(3, 3, 2), 2), predictor_cor = rep(c("No", "Yes"), each = 3))

var_ge_diag <- bgr_nz10 %>%
  filter(cov_str == "diagonal") %>%
  pivot_longer(cols = c(psi_1, psi_5, psi_9), 
               names_to = "parameter", values_to = "estimate") %>%
  ggplot(aes(parameter, estimate)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21) +
  geom_segment(data = df_true, aes(x = parameter-.4, xend = parameter+.4,
                                   y = y_value, yend = y_value,
                                   linetype = "True Parameter"), show.legend = F) +
  facet_wrap(~ predictor_cor, nrow = 2,
             labeller = labeller(predictor_cor = corr.labs)) +
  labs(y = "Estimated parameter", x = TeX("Entry in $\\Psi")) +
  scale_x_discrete(labels = c( TeX("$\\Psi_{1,1}"), TeX("$\\Psi_{2,2}"), TeX("$\\Psi_{3,3}"))) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
  
var_ge_diag

ggsave(
  "../plots/gene/var_ge_diag.pdf",
  var_ge_diag,
  width = 20,
  height = 15,
  units = "cm"
)
```
### Unstructured

```{r}
Lsym <- matrix(c(sqrt(3), 0 ,0, 1, sqrt(3), 0, -sqrt(2), 1, sqrt(2)), nrow = 3, byrow = T)
df_true <- data.frame(parameter = rep(1:9, 2), 
                      y_value = rep(as.vector(Lsym %*% t(Lsym)), 2), predictor_cor = rep(c("No", "Yes"), each = 9))

var_ge_sym <- bgr_nz10 %>%
  filter(cov_str == "unstructured") %>%
  pivot_longer(cols = paste0("psi_", 1:9), 
               names_to = "parameter", values_to = "estimate") %>%
  ggplot(aes(parameter, estimate)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21) +
  geom_segment(data = df_true, aes(x = parameter-.4, xend = parameter+.4,
                                   y = y_value, yend = y_value,
                                   linetype = "True Parameter"), show.legend = FALSE) +
  facet_wrap(~ predictor_cor, nrow = 2,
             labeller = labeller(predictor_cor = corr.labs)) +
  scale_x_discrete(labels = c( TeX("$\\Psi_{1,1}"),
                               TeX("$\\Psi_{1,2}"),
                               TeX("$\\Psi_{1,3}"),
                               TeX("$\\Psi_{2,1}"),
                               TeX("$\\Psi_{2,2}"),
                               TeX("$\\Psi_{2,3}"),
                               TeX("$\\Psi_{3,1}"),
                               TeX("$\\Psi_{3,2}"),
                               TeX("$\\Psi_{3,3}"))) +
  labs(y = "Estimated parameter", x = TeX("Entry in $\\Psi")) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
var_ge_sym
ggsave(
  "../plots/gene/var_ge_sym.pdf",
  var_ge_sym,
  width = 20,
  height = 15,
  units = "cm"
)
```

Combine into final plot

```{r}
(var_ge_scalar + var_ge_diag) / var_ge_sym + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gene/var_ge_comp.pdf", width = 40, height = 30, units = "cm")
```

# GWAS

Check that there are 200 rows for each setting
```{r}
gwas_results %>%
  count(setting)
```

Get results for only the best $\lambda$ for each dataset.
```{r}
best_gwas_results <- gwas_results %>%
  group_by(setting, data_id) %>%
  arrange(bic, .by_group = TRUE) %>%
  filter(row_number() == 1) %>%
  mutate(
    cor = str_extract(setting, "rho([0-9]+\\.[0-9]+)", group = 1),
    penalty = setting %>% 
      str_extract("cov.*_(.*)-", group = 1) %>% toupper %>% factor,
    q = str_extract(setting, "random([0-9])", group = 1),
    cov_str = recode_factor(
      str_extract(setting, "cov(.*)_", group = 1),
      id = "scalar", diag = "diagonal", sym = "unstructured"
    ),
    predictor_cor = if_else(cor == "0.0", "No", "Yes"),
    fp = n_nz - tp,
    f = (2*tp)/(tp+fp+10)
  ) 
```


## Variable selection

### Main text
False positives

```{r}
gwas_main_fp <- best_gwas_results  %>%
  filter(q==3) %>%
  mutate(fpr = fp/990) %>%
  ggplot(aes(x=1, y = fpr, col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(pch=21, position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2)) +
  labs(y = "false positive rate", x = TeX("$\\Psi_\\theta$ structure")) + 
  facet_grid2(. ~ cov_str, scales = "free_y", independent = "y") +
  #scale_y_continuous(expand = expansion(mult = c(0, .05))) +
  theme(legend.position = "top",
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gwas_main_fp

ggsave(
  "../plots/gwas/gwas_main_fp.pdf",
  gwas_main_fp,
  width = 20,
  height = 15,
  units = "cm"
)
```

True positives

```{r}
# Export the data for this plot to a csv for Rosa to make a table

best_gwas_results %>%
  ungroup() %>%
  filter(q==3) %>%
  select(penalty, cov_str, tp) %>%
  write_csv("../sim_results/true_positive_tables/gwas/gwas_main_tp.csv")


```



```{r}
counts <- best_gwas_results %>%
  group_by(penalty, cov_str, q) %>%
  summarise(
    cnt_perfect = sum(tp == 10), .groups = "drop",
    ) %>% 
  filter(q == 3)


gwas_main_tp <- best_gwas_results  %>%
  filter(q==3) %>%
  group_by(cov_str, q) %>%
  ungroup() %>%
  filter(tp < 10) %>%
  ggplot(aes(y = tp, x = cov_str, col = penalty)) +
  geom_point(pch = 21, 
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = .05,
               jitter.width = 0.3)) +
  geom_text(data = counts, aes(cov_str, 10, label = cnt_perfect),
            show.legend = FALSE,
            position = position_dodge(width = 0.9), size = 5) +
  labs(y = "# true positives (out of 10)", x = TeX("$\\Psi_\\theta$ structure")) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 1), 
                     limits = c(0,10), breaks = c(3,6,9,10)) +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gwas_main_tp

ggsave(
  "../plots/gwas/gwas_main_tp.pdf",
  gwas_main_fp,
  width = 20,
  height = 8,
  units = "cm"
)


```

Now let's add them together

```{r}
gwas_main_fp + gwas_main_tp + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gwas/gwas_main_select.pdf", width = 40, height = 10, units = "cm")
```




### Supplementary material

False positives for case $q=1$ and $q=5$:

```{r}

gwas_supp_fp <- best_gwas_results  %>%
  filter(q!=3) %>%
  group_by(penalty, q) %>%
  mutate(outlier = fp > quantile(fp, .75) + IQR(fp) * 1.5) %>%
  ungroup() %>%
  ggplot(aes(y = fp/990, x = q, col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), show.legend = F,
               width = .3, outlier.shape = 21) +
  labs(y = "false positive rate", x = TeX("# random effects ($q$)")) +
  #scale_y_continuous(expand = expansion(mult = c(0, .05))) +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gwas_supp_fp

ggsave(
  "../plots/gwas/gwas_supp_fp.pdf",
  gwas_main_fp,
  width = 20,
  height = 8,
  units = "cm"
)
```
Now true positives 
```{r}
counts <- best_gwas_results %>%
  group_by(penalty, cov_str, q) %>%
  summarise(
    cnt_perfect = sum(tp == 10), .groups = "drop",
    ) %>% 
  filter(q != 3)


gwas_supp_tp <- best_gwas_results  %>%
  filter(q != 3) %>%
  group_by(cov_str, q) %>%
  ungroup() %>%
  filter(tp < 10) %>%
  ggplot(aes(y = tp, x = q, col = penalty)) +
  geom_point(pch = 21, 
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = .05,
               jitter.width = 0.3)) +
  geom_text(data = counts, aes(q, 10, label = cnt_perfect),
            show.legend = FALSE,
            position = position_dodge(width = 0.9), size = 5) +
  labs(y = "# true positives (out of 10)", x = TeX("# random effects ($q$)")) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 1), 
                     limits = c(0,10), breaks = c(1,5,9,10)) +
  theme(legend.position = "top",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()

gwas_supp_tp

ggsave(
  "../plots/gwas/gwas_supp_tp.pdf",
  gwas_supp_tp,
  width = 20,
  height = 8,
  units = "cm"
)
```

Now combine:

```{r}
gwas_supp_fp + gwas_supp_tp + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gwas/gwas_supp_select.pdf", width = 40, height = 10, units = "cm")
```




## Estimation



```{r}
df_true <- data.frame(y_value = 4)
gwas_main_beta3 <- best_gwas_results %>%
  filter(q == 3) %>%
  group_by(penalty, cov_str) %>%
  mutate(outlier = beta_3 > quantile(beta_3, .75) + IQR(beta_3) * 1.5 | 
           beta_3 < quantile(beta_3, .25) - IQR(beta_3) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_3, x = cov_str, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_3$"), x = TeX("$\\Psi_\\theta$ structure")) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
}
gwas_main_beta3
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gwas/gwas_main_beta3.pdf",
  gwas_main_beta3,
  width = 20,
  height = 8,
  units = "cm"
)
```
```{r}
df_true <- data.frame(y_value = 3)
gwas_main_beta4 <- best_gwas_results %>%
  filter(q == 3) %>%
  group_by(penalty, cov_str) %>%
  mutate(outlier = beta_4 > quantile(beta_4, .75) + IQR(beta_4) * 1.5 | 
           beta_4 < quantile(beta_4, .25) - IQR(beta_4) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_4, x = cov_str, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_4$"), x = TeX("$\\Psi_\\theta$ structure")) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
}
gwas_main_beta4
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gwas/gwas_main_beta4.pdf",
  gwas_main_beta4,
  width = 20,
  height = 8,
  units = "cm"
)
```


Combine

```{r}
gwas_main_beta3 + gwas_main_beta4 + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gwas/gwas_main_est.pdf", width = 40, height = 10, units = "cm")

```


```{r}
df_true <- data.frame(y_value = 4)
gwas_supp_beta3 <- best_gwas_results %>%
  filter(q != 3) %>%
  group_by(penalty, q) %>%
  mutate(outlier = beta_3 > quantile(beta_3, .75) + IQR(beta_3) * 1.5 | 
           beta_3 < quantile(beta_3, .25) - IQR(beta_3) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_3, x = q, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_3$"), x = TeX("# random effects ($q$)")) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
}
gwas_supp_beta3
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gwas/gwas_supp_beta3.pdf",
  gwas_supp_beta3,
  width = 20,
  height = 8,
  units = "cm"
)
```

```{r}
df_true <- data.frame(y_value = -3)
gwas_supp_beta8 <- best_gwas_results %>%
  filter(q != 3) %>%
  group_by(penalty, q) %>%
  mutate(outlier = beta_8 > quantile(beta_8, .75) + IQR(beta_8) * 1.5 | 
           beta_8 < quantile(beta_8, .25) - IQR(beta_8) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_8, x = q, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = TeX("$\\beta_8$"), x = TeX("# random effects ($q$)")) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
}
gwas_supp_beta8
#Note that there are fewer than 100 data-sets for some of the lasso settings (but none of the scad settings)!
ggsave(
  "../plots/gwas/gwas_supp_beta8.pdf",
  gwas_supp_beta8,
  width = 20,
  height = 8,
  units = "cm"
)
```
Combine 

```{r}
gwas_supp_beta3 + gwas_supp_beta8 + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gwas/gwas_supp_est.pdf", width = 40, height = 10, units = "cm")

```

Let's get MC means and sds of beta_8 estimates in all GWAS settings with SCAD penalty:
```{r}
gwas_beta8 <- best_gwas_results %>% 
  filter(penalty == "SCAD") %>%
  select(setting, data_id, beta_8) %>%
  mutate(identified = beta_8!=0) %>%
    filter(identified) %>%
  group_by(setting) %>%
  summarise(recovered = n(), mean = mean(beta_8), 
            bias = mean + 3, sd = sd(beta_8))

gwas_beta8
```

## Variance components

### Scalar variance components

```{r}
df_true <- data.frame(y_value = 0.56) 

var_gwas_scalar <- best_gwas_results %>%
  filter(cov_str == "scalar") %>%
  ggplot(aes(y = psi_1, x=q)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21)  +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  labs(y = "Estimated parameter", x = TeX("# random effects ($q$)")) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
  
var_gwas_scalar

ggsave(
  "../plots/gwas/var_gwas_scalar.pdf",
  var_gwas_scalar,
  width = 20,
  height = 15,
  units = "cm"
)
```


### Diagonal variance components

If we want to check these, we must divide the elements of $\Psi$ by the variances of the corresponding columns of $Z$ because right now, they are on the scale of the standardized covariates (when we ran GWAS simulations, we had a previous version of the code that does not convert the random effects matrix back to the original scale).

As such, we have to read in the GWAS data matrices
```{r, cache=TRUE, message=FALSE}
idxs = 1:100
diag_dfs <- idxs %>% map( 
                \(i) read_csv(paste0("../data/GWAS/random3_covdiag/data", i, ".csv"))
                )

variances <- diag_dfs %>% 
  map(\(df) tibble(x2_var = var(df$X2), x3_var = var(df$X3))) %>%
  list_rbind 
variances <- variances %>%
  mutate(data_id = paste0("data", row_number()))

```


```{r}
df_true <- data.frame(parameter = rep(c(1, 2, 3), 2), 
                      y_value = rep(c(3, 3, 2), 2), correlation = rep(c("No", "Yes"), each = 3))

var_gwas_diag <- best_gwas_results %>%
  filter(cov_str == "diagonal") %>%
  left_join(variances, by = "data_id") %>%
  mutate(true_psi1 = psi_1, true_psi5 = psi_5/x2_var, true_psi9 = psi_9/x3_var) %>%
  pivot_longer(cols = c(true_psi1, true_psi5, true_psi9), 
               names_to = "parameter", values_to = "estimate") %>%
  ggplot(aes(parameter, estimate)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21) +
  geom_segment(data = df_true, aes(x = parameter-.4, xend = parameter+.4,
                                   y = y_value, yend = y_value,
                                   linetype = "True Parameter"),
               show.legend = FALSE) +
  labs(y = "Estimated parameter", x = TeX("Diagonal $\\Psi")) +
  scale_x_discrete(labels = c( TeX("$\\Psi_{1,1}"), TeX("$\\Psi_{2,2}"), TeX("$\\Psi_{3,3}"))) +

  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
  
var_gwas_diag

ggsave(
  "../plots/gwas/var_gwas_diag.pdf",
  var_gwas_diag,
  width = 20,
  height = 30,
  units = "cm"
)
```
### Unstructured

```{r, cache=TRUE, message=FALSE}
idxs = 1:100
sym_dfs <- idxs %>% map( 
                \(i) read_csv(paste0("../data/GWAS/random3_covsym/data", i, ".csv"))
                )

sds <- sym_dfs %>% 
  map(\(df) tibble(x2_sd = sd(df$X2), x3_sd = sd(df$X3))) %>%
  list_rbind 
sds <- sds %>%
  mutate(data_id = paste0("data", row_number()))

```


```{r}

Lsym <- matrix(c(sqrt(3), 0 ,0, 1, sqrt(3), 0, -sqrt(2), 1, sqrt(2)), nrow = 3, byrow = T)
df_true <- data.frame(parameter = rep(1:9, 2), 
                      y_value = rep(as.vector(Lsym %*% t(Lsym)), 2))

var_gwas_sym <- best_gwas_results %>%
  filter(cov_str == "unstructured") %>%
  left_join(sds, by = "data_id") %>%
  mutate(
    true_psi1 = psi_1, 
    true_psi2 = psi_2/x2_sd, 
    true_psi3 = psi_3/x3_sd,
    true_psi4 = psi_4/x2_sd,
    true_psi5 = psi_5/x2_sd^2,
    true_psi6 = psi_6/(x2_sd*x3_sd),
    true_psi7 = psi_7/x3_sd,
    true_psi8 = psi_8/(x2_sd*x3_sd),
    true_psi9 = psi_9/x3_sd^2
    ) %>%
  pivot_longer(cols = paste0("true_psi", 1:9), 
               names_to = "parameter", values_to = "estimate") %>%
  ggplot(aes(parameter, estimate)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21) +
  geom_segment(data = df_true, aes(x = parameter-.4, xend = parameter+.4,
                                   y = y_value, yend = y_value,
                                   linetype = "True Parameter"),
               show.legend = FALSE) +
  scale_x_discrete(labels = c( TeX("$\\Psi_{1,1}"),
                               TeX("$\\Psi_{1,2}"),
                               TeX("$\\Psi_{1,3}"),
                               TeX("$\\Psi_{2,1}"),
                               TeX("$\\Psi_{2,2}"),
                               TeX("$\\Psi_{2,3}"),
                               TeX("$\\Psi_{3,1}"),
                               TeX("$\\Psi_{3,2}"),
                               TeX("$\\Psi_{3,3}"))) +
  labs(y = "Estimated parameter", x = TeX("Unstructured symmetric $\\Psi")) + 
    theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
var_gwas_sym

ggsave(
  "../plots/gwas/var_gwas_sym.pdf",
  var_gwas_sym,
  width = 20,
  height = 15,
  units = "cm"
)
```
Combine into final plot

```{r}
(var_gwas_scalar + var_gwas_diag) / var_gwas_sym + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/gwas/var_gwas_comp.pdf", width = 40, height = 30, units = "cm")
```

# Microbiome

Check that there are 1200 (2 best lambdas * 6 otu-covariance structures * 100 datasets) rows for each random effect setting
```{r}
otu_results %>%
  count(setting)
```

Get results for only the best $\lambda$ for each dataset.
```{r}
best_otu_results <- otu_results %>%
  group_by(setting, data_id) %>%
  arrange(bic, .by_group = TRUE) %>%
  filter(row_number() == 1) %>%
  mutate(
    otu_cor = str_extract(data_id, "_(.*)_", group = 1),
    penalty = setting %>% 
      str_extract("cov.*_(.*)-", group = 1) %>% toupper %>% factor(levels = c("LASSO", "SCAD")),
    q = str_extract(setting, "random([0-9])", group = 1),
    cov_str = str_extract(setting, "cov(.*)_", group = 1) %>% 
      case_match( 
        "id" ~ "scalar",
        "diag" ~ "diagonal",
        "sym" ~ "unstructured"
      ) %>% 
      factor(levels =  c("scalar", "diagonal", "unstructured")), 
    fp = n_nz - tp,
    f = (2*tp)/(tp+fp+11),
    beta_ind2 = if_else(q == 1, beta_3, beta_2),
    beta_grp = if_else(q == 1, beta_2, if_else(q == 3, beta_4, beta_6))
  ) 

```

## Variable selection

### Main text

```{r}
otu_main_fp <- best_otu_results  %>%
  filter(q==3, otu_cor %in% c("band", "scale_free")) %>%
  ggplot(aes(y = fp/117, x = 1, col = otu_cor)) +
  geom_boxplot(outlier.shape = NA, show.legend = F, width = .3) +
   geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitter(height = 0, width = .1)) +
  facet_grid2(otu_cor ~ cov_str, scales = "free_y",
             labeller = labeller(predictor_cor = corr.labs)) +
  labs(y = "false positive rate", x = TeX("$\\Psi_\\theta$ structure")) + 
  theme(legend.position = "none",
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        text = element_text(size = 18)) +
  #scale_y_continuous(expand = expansion(mult = c(0, .05))) +
  scale_color_otu()
  

otu_main_fp
ggsave(
  "../plots/otu/otu_main_fp.pdf",
  otu_main_fp,
  width = 20,
  height = 15,
  units = "cm"
)
```
True positives

```{r}
#Export to csv for Rosa to make a table
best_otu_results %>%
  ungroup() %>%
  filter(q==3) %>%
  select(q, otu_cor, cov_str, tp) %>%
  write_csv("../sim_results/true_positive_tables/otu/otu_main_tp.csv")
```



```{r}
counts <- best_otu_results %>%
  group_by(otu_cor, cov_str, q, penalty) %>%
  summarise(
    cnt_perfect = sum(tp == 11), .groups = "drop",
    ) %>% 
  filter(q == 3, otu_cor %in% c("band", "scale_free"))

otu_main_tp <- best_otu_results  %>%
  filter(q == 3, otu_cor %in% c("band", "scale_free")) %>%
  filter(tp < 11) %>%
  ggplot(aes(y = tp, x = cov_str, col = otu_cor)) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitter(height = 0, width = 0.1)) +
  geom_text(data = counts, aes(cov_str, 11, label = cnt_perfect),
            show.legend = FALSE, size = 5) +
  facet_wrap(~ otu_cor, nrow = 2) +
  labs(y = "# true positives (out of 11)", x = TeX("$\\Psi_\\theta$ structure")) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 1), 
                     limits = c(0,12), breaks = c(4,11)) +
  theme(legend.position = "none",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_otu()

otu_main_tp

ggsave(
  "../plots/otu/otu_main_tp.pdf",
  otu_main_tp,
  width = 20,
  height = 15,
  units = "cm"
)

```

Combine false and true positives

```{r}
otu_main_fp + otu_main_tp + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') 
ggsave("../plots/otu/otus_main_select.pdf", width = 40, height = 10, units = "cm")

```


### Supplementary material

False positives


Band and scale free
```{r}
otu_supp_fp_bsf <- best_otu_results  %>%
  filter(q!=3, otu_cor %in% c("band", "scale_free")) %>%
  ggplot(aes(y = fp/117, x = q, col = otu_cor)) +
  geom_boxplot(width = .3, outlier.shape = NA, show.legend = F) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitter(height = 0, width = 0.1)) +
  facet_wrap(~ otu_cor, nrow = 2) +
  labs(y = "false positive rate", 
       x = TeX("# random effects ($q$)"),
       color = "network") + 
  theme(legend.position = "none",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  #scale_y_continuous(expand = expansion(mult = c(0, .05))) +
  scale_color_otu()

otu_supp_fp_bsf
ggsave(
  "../plots/otu/otu_supp_fp_bsf.pdf",
  otu_supp_fp_bsf,
  width = 20,
  height = 10,
  units = "cm"
)
```

Other OTU correlation structures

```{r}
q.labs = c("q: 1", "q: 3", "q: 5")
names(q.labs) <- c("1", "3", "5")
otu_supp_fp_others <- best_otu_results  %>%
  filter( !(otu_cor %in% c("band", "scale_free"))) %>%
  group_by(q, otu_cor, cov_str) %>%
  mutate(outlier = fp > quantile(fp, .75) + IQR(fp) * 1.5) %>%
  ungroup() %>% {
  ggplot(data = ., aes(y = fp/117, x = cov_str, col = penalty)) +
  geom_boxplot(width = .3, outlier.shape = NA, show.legend = F) +
   geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitter(height = 0, width = 0.1)) +
  facet_grid(otu_cor ~ q, space = "free", scales = "free_x",
             labeller = labeller(q = q.labs)) +  
  labs(y = "false positive rate", x = TeX("$\\Psi_\\theta$ structure")) + 
  theme(legend.position = "none",
        axis.ticks.x = element_blank(),
        strip.text.y = element_text(size = 8),
        text = element_text(size = 18)) +
  scale_color_mine() +
  scale_y_continuous(n.breaks = 4)
  }

otu_supp_fp_others
ggsave(
  "../plots/otu/otu_supp_fp_others.pdf",
  otu_supp_fp_bsf,
  width = 20,
  height = 10,
  units = "cm"
)
```

True positives

Band and scale free

```{r}
counts <- best_otu_results %>%
  group_by(otu_cor, cov_str, q, penalty) %>%
  summarise(
    cnt_perfect = sum(tp == 11), .groups = "drop",
    ) %>% 
  filter(q != 3, otu_cor %in% c("band", "scale_free"))

otu_supp_tp_bsf <- best_otu_results  %>%
  filter(q != 3, otu_cor %in% c("band", "scale_free")) %>%
  filter(tp < 11) %>%
  ggplot(aes(y = tp, x = q, col = otu_cor)) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitter(height = 0, width = 0.1)) +
  geom_text(data = counts, aes(q, 11, label = cnt_perfect),
            show.legend = FALSE, size = 5) +
  facet_wrap(~ otu_cor, nrow = 2) +
  labs(y = "# true positives (out of 11)", 
       x = TeX("# random effects ($q$)"),
       color = "network") + 
  scale_y_continuous(labels = scales::label_number(accuracy = 1), 
                     limits = c(0,12), breaks = c(6,11)) +
  theme(legend.position = "none",
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_otu()


otu_supp_tp_bsf
ggsave(
  "../plots/otu/otu_supp_tp_bsf.pdf",
  otu_supp_fp_bsf,
  width = 20,
  height = 10,
  units = "cm"
)
```
Other OTU correlation structures

```{r}
counts <- best_otu_results %>%
  group_by(otu_cor, cov_str, q, penalty) %>%
  summarise(
    cnt_perfect = sum(tp == 11), .groups = "drop",
    ) %>% 
  filter(!(otu_cor %in% c("band", "scale_free")))

otu_supp_tp_others <- best_otu_results  %>%
  filter( !(otu_cor %in% c("band", "scale_free"))) %>%
  filter(tp < 11) %>%
  ggplot(aes(y = tp, x = cov_str, col = penalty)) +
  geom_point(pch = 21, #aes(alpha = plot),
             position = position_jitter(height = 0, width = 0.1)) +
  geom_text(data = counts, aes(cov_str, 11, label = cnt_perfect),
            show.legend = FALSE, size = 5) +
  facet_grid(otu_cor ~ q, space = "free", scales = "free_x",
             labeller = labeller(q = q.labs)) +  
  labs(y = "# true positives (out of 11)", x = TeX("$\\Psi_\\theta$ structure")) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 1), 
                     limits = c(0,12), breaks = c(4,6,11)) + 
  theme(legend.position = "none",
        axis.text.y = element_text(size = 12),
        axis.ticks.x = element_blank(),
        strip.text.y = element_text(size = 8),
        text = element_text(size = 18)) +
  scale_color_mine()

otu_supp_tp_others
ggsave(
  "../plots/otu/otu_supp_tp_others.pdf",
  otu_supp_fp_bsf,
  width = 20,
  height = 10,
  units = "cm"
)
```
Combine into single plot

```{r}
otu_supp_fp_bsf + otu_supp_tp_bsf + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A') & theme(legend.position = 'bottom') 
ggsave("../plots/otu/otu_supp_select_bsf.pdf", width = 40, height = 15, units = "cm")
```


```{r}
otu_supp_fp_others + otu_supp_tp_others + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'A')
ggsave("../plots/otu/otu_supp_select_others.pdf", width = 40, height = 15, units = "cm")
```

## Estimation

Let's look at the coefficient on the group-level measured variable:

$q=3$ with band and scale-free OTU-OTU covariance matrices:
```{r}
df_true <- data.frame(y_value = -1)
otu_main_betagrp <- best_otu_results %>%
  filter(q==3, otu_cor %in% c("band", "scale_free")) %>%
  group_by(otu_cor, cov_str) %>%
  mutate(outlier = beta_grp > quantile(beta_grp, .75) + IQR(beta_grp) * 1.5 | 
           beta_grp < quantile(beta_grp, .25) - IQR(beta_grp) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_grp, x = cov_str, 
                      col = otu_cor)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  facet_wrap(~ otu_cor, nrow = 2) +
  labs(y = TeX("$\\beta_{group}$"), x = TeX("$\\Psi_\\theta$ structure")) +
  coord_cartesian(ylim = c(-2, 0)) +
  theme(legend.position = "none",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_otu()
}
otu_main_betagrp

ggsave(
  "../plots/otu/otu_main_betagrp.pdf",
  otu_main_betagrp,
  width = 20,
  height = 10,
  units = "cm"
)
```

$q=1$ or $5$ with band and scale-free OTU-OTU covariance matrices:
```{r}
df_true <- data.frame(y_value = -1)
otu_supp_betagrp_bsf <- best_otu_results %>%
  filter(q!=3, otu_cor %in% c("band", "scale_free")) %>%
  group_by(otu_cor, q) %>%
  mutate(outlier = beta_grp > quantile(beta_grp, .75) + IQR(beta_grp) * 1.5 | 
           beta_grp < quantile(beta_grp, .25) - IQR(beta_grp) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_grp, x = q, 
                      col = otu_cor)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  facet_wrap(~ otu_cor, nrow = 2) +
  labs(y = TeX("$\\beta_{group}$"), x = TeX("# random effects ($q$)")) +
  coord_cartesian(ylim = c(-2, 0)) +
  theme(legend.position = "none",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_otu()
}
otu_supp_betagrp_bsf

ggsave(
  "../plots/otu/otu_supp_betagrp_bsf.pdf",
  otu_supp_betagrp_bsf,
  width = 20,
  height = 10,
  units = "cm"
)
```

Other OTU-correlation structures:
```{r}
df_true <- data.frame(y_value = -1)
otu_supp_betagrp_other <- best_otu_results %>%
  filter(!(otu_cor %in% c("band", "scale_free"))) %>%
  group_by(otu_cor, q, cov_str) %>%
  mutate(outlier = beta_grp > quantile(beta_grp, .75) + IQR(beta_grp) * 1.5 | 
           beta_grp < quantile(beta_grp, .25) - IQR(beta_grp) * 1.5) %>%
  ungroup() %>%
  {
  ggplot(data =., aes(y = beta_grp, x = cov_str, 
                      col = penalty)) +
  geom_boxplot(position = position_dodge(width = 0.9), 
               outlier.shape = NA) +
  geom_point(data = filter(., outlier), 
             pch = 21, #aes(alpha = plot),
             position = position_jitterdodge(
               dodge.width = 0.9,
               jitter.height = 0,
               jitter.width = 0.2), show.legend  = F) +
  geom_hline(show.legend = FALSE,data = df_true, aes(yintercept = y_value, linetype = "True Parameter")) +
  facet_grid(otu_cor ~ q, space = "free", scales = "free_x",
             labeller = labeller(q = q.labs)) +  
  labs(y = TeX("$\\beta_{group}$"), x = TeX("$\\Psi_\\theta$ structure")) +
  coord_cartesian(ylim = c(-2, 0)) +
  theme(legend.position = "none",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        strip.text.y = element_text(size = 8),
        text = element_text(size = 18)) +
  scale_color_mine()
}
otu_supp_betagrp_other

ggsave(
  "../plots/otu/otu_supp_betagrp_other.pdf",
  otu_supp_betagrp_other,
  width = 20,
  height = 10,
  units = "cm"
)
```

Let's get MC means and sds of beta_8 estimates in all OTU settings:
```{r}
otu_beta8 <- best_otu_results %>% 
  select(setting, otu_cor, data_id, beta_9) %>%
  mutate(identified = beta_9!=0) %>%
  filter(identified) %>%
  group_by(setting, otu_cor) %>%
  summarise(recovered = n(), mean = mean(beta_9), 
            bias = mean + 3, sd = sd(beta_9))

otu_beta8
```


## Variance components

```{r}
df_true <- data.frame(parameter = rep(c(1, 2, 3), 6), 
                      y_value = rep(c(.56, 1, 1), 6), 
                      correlation = rep(unique(best_otu_results$otu_cor), 
                                        each = 3))

var_gwas_diag <- best_otu_results %>%
  filter(cov_str == "diagonal") %>%
  pivot_longer(cols = c(psi_1, psi_5, psi_9), 
               names_to = "parameter", values_to = "estimate") %>%
  ggplot(aes(parameter, estimate)) +
  geom_boxplot(position = position_dodge(width = 0.9), aes(col = penalty), outlier.shape = 21) +
  coord_cartesian(ylim = c(0,2)) +
  geom_segment(data = df_true, aes(x = parameter-.4, xend = parameter+.4,
                                   y = y_value, yend = y_value,
                                   linetype = "True Parameter"),
               show.legend = FALSE) +
  facet_wrap(~ otu_cor) +
  labs(y = "Estimated parameter", x = TeX("Diagonal $\\Psi")) +
  scale_x_discrete(labels = c( TeX("$\\Psi_{1,1}"), TeX("$\\Psi_{2,2}"), TeX("$\\Psi_{3,3}"))) +
  theme(legend.position = "top",
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18)) +
  scale_color_mine()
var_gwas_diag
```


# Inter-omics comparison

```{r}
beta8_results <- bind_rows(list(gene = gene_beta8, 
                                gwas = gwas_beta8, 
                                otu = otu_beta8), .id = "dtype") %>%
  mutate(cov_str = recode_factor(
      str_extract(setting, "cov(.*)_", group = 1),
      id = "scalar", diag = "diagonal", sym = "unstructured"
      ),
    q = str_c("q=", str_extract(setting, "random([0-9])", group = 1)),
    cov_str = factor(fct_cross(q, cov_str),
                     levels = c("q=1:scalar", "q=3:scalar", "q=3:diagonal", 
                                "q=3:unstructured", "q=5:scalar")), 
    cor_features = str_extract(setting, "rho([0-9]+\\.[0-9]+)", group = 1),
    cor_features = if_else(cor_features == "0.0", "uncorrelated", "correlated")) %>%
  mutate(grouping = case_when(
    dtype == "otu" ~ paste0("otu:", otu_cor),
    dtype == "gene" ~ paste0("gene:", cor_features),
    dtype == "gwas" ~ "gwas"
  ))

```



```{r}
ggplot(beta8_results, aes(cov_str, sd, color = dtype)) +
  geom_line(aes(group = grouping)) +
  geom_point() +
  labs(color = "Omics type",
       x = "Random effect structure", 
       y = TeX("$SD(\\hat{\\beta_8} | \\hat{\\beta_8} \\neq 0)$")) +
  expand_limits(x = c(NA, 6)) +
  geom_dl(aes(label=grouping), method = list("last.points", dl.trans(x=x+.2), cex=0.7))

ggplot(beta8_results, aes(cov_str, 100-recovered, color = dtype)) +
  geom_line(aes( group = grouping)) +
  geom_point() +
  labs(color = "Omics type", 
       x = "Random effect structure", 
       y = TeX("# times $\\beta_8$ incorrectly estimated zero$")) + 
  expand_limits(x = c(NA, 6)) +
  geom_dl(aes(label=grouping), method = list("last.points", dl.trans(x=x+.2), cex=.7))
```






